### PR TITLE
WIP: Data-Dictionary-FieldWidget

### PR DESCRIPTION
**Pull request by kaise-lafrai**
New Data-Dictionary-Widget created for the purposes of displaying fields for the field_json_metadata data node to collect information for the data-dictionary

fixes [org/repo/issue#]

- [ ] Test coverage exists
- [ ] Documentation exists

## QA Steps

- [ ] Add manual QA steps in checklist format for a reviewer to perform to confirm that the feature or fix is working. Include as much details as possible so that the reviewer doesn't lose time figuring out how to perform steps.

### CODE REVIEW

File: modules/data_dictionary_widget/data_dictionary_widget.info.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# custom_widget.info.yml
```
Comment (kaise-lafrai) [inline modules/data_dictionary_widget/data_dictionary_widget.info.yml]: Remember to remove this

File: modules/data_dictionary_widget/data_dictionary_widget.info.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+name: 'Data Dictionary Widget'
+description: 'Provides a field widget to generate a data-dictionary form'
+package: DKAN
+type: module
+core_version_requirement: ^9.4 || ^10
+dependencies:
+  - drupal:field
```
Comment (kaise-lafrai) [inline modules/data_dictionary_widget/data_dictionary_widget.info.yml]: Add return line after this row

File: modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace Drupal\data_dictionary_widget\Plugin\Field\FieldWidget;
+
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Entity\FieldableEntityInterface;
+use Drupal\Core\Field\WidgetBase;
+use Drupal\Core\Form\FormStateInterface;
+
+/**
+ * A widget bar.
```
Comment (kaise-lafrai) [inline modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php]: Update this

File: modules/data_dictionary_widget/src/Controller/DataDictionary.php [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      $exsisting_identifiers[$node->id()] .= $node->uuid();
+    }
+
+    return $exsisting_identifiers;
+}
+
+
+public static function setAjaxElements(array $dictionaryFields){
+  foreach ($dictionaryFields['data']['#rows'] as $row => $data) {
+    $edit_button = $dictionaryFields['edit_buttons'][$row];
```
Comment (janette) [inline modules/data_dictionary_widget/src/Controller/DataDictionary.php]:
```
$edit_button = $dictionaryFields['edit_buttons'][$row] ?? NULL;
$edit_fields = $dictionaryFields['edit_fields'][$row] ?? NULL;
```

File: modules/data_dictionary_widget/src/Controller/DataDictionary.php [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+          <li><b>email</b>: A valid email address.
+          <li><b>uri</b>: A valid URI.
+          <li><b>binary</b>: A base64 encoded string representing binary data.
+          <li><b>uuid</b>: A string that is a UUID.
+        </ul>";
+    }
+
+    if ($dataType === 'date') {
+      $description .= "
+        <ul>
```
Comment (janette) [inline modules/data_dictionary_widget/src/Controller/DataDictionary.php:81]:
```
        <ul>
          <li><b>default</b>: An ISO8601 format string of YYYY-MM-DD.</li>
          <li><b>any</b>: Any parsable representation of a date. The implementing library can attept to parse the datetime via a range of strategies.</li>
          <li><b>other</b>: If your date values follow a collective but non-ISO8601 pattern, select this option and define the incoming format using the syntax of <a href='https://strftime.org/'>C / Python strftime</a>.
            For example, if your data had dates formatted as MM/DD/YYYY, you would enter %m/%d/%Y into the Other format field.</li>
        </ul>";
```

File: modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+    'format',
+    'edit',
+    'update',
+    'abort',
+    'delete'
+   ];
+
+    //Creating ajax buttons/fields to be placed in correct location later.
+    foreach ($data_results as $key => $data) {
+      if (in_array($op_index[0],$action_list) && array_key_exists($key,  $fields_being_modified)){
```
Comment (janette) [inline modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php:85]: Trying to access array offset on value of type null
```
if (isset($op_index[0]) && in_array($op_index[0], $action_list) && array_key_exists($key, $fields_being_modified)) {
```

File: modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+            '#suffix' => '</div>',
+            '#validated' => TRUE,
+            '#options' => DataDictionary::setFormatOptions($current_fields[$key]['type']),
+          ];
+          $element['dictionary_fields']['edit_fields'][$key]['format_other'] = [
+            '#name' => 'field_json_metadata[0][dictionary_fields][data][' . $key .'][field_collection][format_other]',
+            '#type' => 'textfield',
+            '#title' => $this->t('Other format'),
+            //'#required' => TRUE,
+            '#value' =>  $current_fields[$key]['format_other'],
```
Comment (janette) [inline modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php:138]: '#value' => $current_fields[$key]['format_other'] ?? NULL,

File: modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+ */
+class DataDictionaryWidget extends WidgetBase implements TrustedCallbackInterface {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
+    $field_values = $form_state->getValue(["field_json_metadata"]);
+    $current_fields = $form_state->get('current_fields');
+    $fields_being_modified = $form_state->get("fields_being_modified");
```
Comment (janette) [inline modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php:32]: `$fields_being_modified = $form_state->get("fields_being_modified") ?? NULL;`

File: modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+    $op = $form_state->getTriggeringElement()['#op'] ?? null;
+    $field_json_metadata = !empty($items[0]->value) ? json_decode($items[0]->value, true) : [];
+    $op_index = $form_state->getTriggeringElement()['#op'] ? explode("_", $form_state->getTriggeringElement()['#op']) : null;
```
Comment (janette) [inline modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php:35]: `$op_index = isset($form_state->getTriggeringElement()['#op']) ? explode("_", $form_state->getTriggeringElement()['#op']) : NULL;`

**Review by janette — COMMENTED** [review]
Great progress, just need to check for null values to avoid errors. Also I know this is WIP but good to get the indentation and spacing correct as you go.

File: modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+    // Creating ajax buttons/fields to be placed in correct location later.
+    foreach ($data_results as $key => $data) {
+      if (isset($op_index[0]) && in_array($op_index[0], $action_list) && array_key_exists($key, $fields_being_modified)) {
+        $element['dictionary_fields']['edit_fields'][$key]['name'] = [
+          '#name' => 'field_json_metadata[0][dictionary_fields][data][' . $key . '][field_collection][name]',
+          '#type' => 'textfield',
+          '#value' => $this->t($current_fields[$key]['name']),
+          '#required' => TRUE,
+          '#title' => 'Name',
+          '#description' => $this->t('A name for this field.'),
```
Comment (janette) [inline modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php]: '#description' => $this->t('Machine name of the field/column in the data table.'),

**Comment by janette** [comment]
@tiffneybare one more definition update for **generateFormatDescription**
```
$description = "<p>The format of the data in this field. Supported formats depend on the specified field type:</p>";
```

**Pull request closed without merging**